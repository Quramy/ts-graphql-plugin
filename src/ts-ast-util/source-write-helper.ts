import path from 'path';
import ts from 'typescript';

import { SourceWriteHelper } from './types';
import { isImportDeclarationWithCondition, mergeImportDeclarationsWithSameModules } from './utilily-functions';

const printer = ts.createPrinter();

/**
 *
 * @internal
 *
 */
export class Helper implements SourceWriteHelper {
  private _unmergedComment: string[] = [];
  private _statements: ts.Statement[] = [];

  constructor(private _filename: string) {}

  get outputFileName() {
    return this._filename;
  }

  get outputDirName() {
    return path.dirname(this._filename);
  }

  findImportDeclarationIndex({ isDefault, name, from }: { isDefault?: boolean; name?: string; from?: string }) {
    return this._statements.findIndex(st => isImportDeclarationWithCondition(st, { isDefault, name, from }));
  }

  pushNamedImportIfNeeded(identifierName: string, from: string) {
    if (this.findImportDeclarationIndex({ name: identifierName, from }) !== -1) return false;
    const importSpecifier = ts.createImportSpecifier(undefined, ts.createIdentifier(identifierName));
    const namedBinding = ts.createNamedImports([importSpecifier]);
    const importClause = ts.createImportClause(undefined, namedBinding);
    const moduleSpecifier = ts.createStringLiteral(from);
    const importDeclaration = ts.createImportDeclaration(undefined, undefined, importClause, moduleSpecifier);
    const indexOfSameModuleImport = this.findImportDeclarationIndex({ from });
    if (indexOfSameModuleImport === -1) {
      this.pushImportDeclaration(importDeclaration);
    } else {
      const base = this.getStatements()[indexOfSameModuleImport] as ts.ImportDeclaration;
      const merged = mergeImportDeclarationsWithSameModules(base, importDeclaration);
      this._replaceStatement(indexOfSameModuleImport, merged);
    }
    return true;
  }

  pushDefaultImportIfNeeded(identifierName: string, from: string) {
    if (this.findImportDeclarationIndex({ name: identifierName, from, isDefault: true }) !== -1) return false;
    const importClause = ts.createImportClause(ts.createIdentifier(identifierName), undefined);
    const moduleSpecifier = ts.createStringLiteral(from);
    const importDeclaration = ts.createImportDeclaration(undefined, undefined, importClause, moduleSpecifier);
    const indexOfSameModuleImport = this.findImportDeclarationIndex({ from });
    if (indexOfSameModuleImport === -1) {
      this.pushImportDeclaration(importDeclaration);
    } else {
      const base = this.getStatements()[indexOfSameModuleImport] as ts.ImportDeclaration;
      const merged = mergeImportDeclarationsWithSameModules(base, importDeclaration);
      this._replaceStatement(indexOfSameModuleImport, merged);
    }
    return true;
  }

  pushImportDeclaration(declaration: ts.ImportDeclaration) {
    let i = this._statements.length - 1;
    for (; i >= 0; i--) {
      if (ts.isImportDeclaration(this._statements[i])) break;
    }
    this._insertStatement(i, declaration);
  }

  pushStatement(statement: ts.Statement) {
    this._insertStatement(this._statements.length, statement);
  }

  writeLeadingComment(comment: string) {
    if (this._statements.length > 0) {
      ts.addSyntheticLeadingComment(this._statements[0], ts.SyntaxKind.MultiLineCommentTrivia, ` ${comment} `, true);
    } else {
      this._unmergedComment.push(comment);
    }
  }

  getStatements() {
    return this._statements as ReadonlyArray<ts.Statement>;
  }

  toSourceFile() {
    const sourceFile = ts.createSourceFile(this._filename, '', ts.ScriptTarget.Latest, false, ts.ScriptKind.TS);
    return ts.updateSourceFileNode(sourceFile, this._statements);
  }

  toFileContent() {
    const content = printer.printFile(this.toSourceFile());
    return { fileName: this._filename, content };
  }

  private _insertStatement(index: number, statement: ts.Statement) {
    this._statements = [...this._statements.slice(0, index), statement, ...this._statements.slice(index)];
    if (this._unmergedComment.length) {
      this._unmergedComment.forEach(comment => this.writeLeadingComment(comment));
      this._unmergedComment = [];
    }
  }

  private _replaceStatement(index: number, statement: ts.Statement) {
    if (!this._statements[index]) return;
    this._statements[index] = statement;
  }
}

export function createSourceWriteHelper({ outputFileName }: { outputFileName: string }) {
  return new Helper(outputFileName);
}
