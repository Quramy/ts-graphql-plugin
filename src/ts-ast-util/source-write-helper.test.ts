import ts from 'typescript';
import { createSourceWriteHelper, Helper } from './source-write-helper';

describe(createSourceWriteHelper, () => {
  describe(Helper.prototype.pushStatement, () => {
    it('should push statement', () => {
      const helper = createSourceWriteHelper({ outputFileName: 'out.ts' });
      helper.pushStatement(ts.createExpressionStatement(ts.createIdentifier('hoge')));
      expect(helper.getStatements().length).toBe(1);
    });
  });
  describe(Helper.prototype.replaceStatement, () => {
    it('should replace exsiting statement to new statement', () => {
      const helper = createSourceWriteHelper({ outputFileName: 'out.ts' });
      const st = ts.createExpressionStatement(ts.createIdentifier('hoge'));
      const newSt = ts.createExpressionStatement(ts.createIdentifier('foo'));
      helper.pushStatement(st);
      expect(helper.replaceStatement(st, newSt)).toBeTruthy();
      expect(helper.toFileContent().content.trim()).toBe('foo;');
    });

    it('should return false when the statement is not foundt', () => {
      const helper = createSourceWriteHelper({ outputFileName: 'out.ts' });
      const st = ts.createExpressionStatement(ts.createIdentifier('hoge'));
      const newSt = ts.createExpressionStatement(ts.createIdentifier('foo'));
      const st2 = ts.createExpressionStatement(ts.createIdentifier('bar'));
      helper.pushStatement(st2);
      expect(helper.replaceStatement(st, newSt)).toBeFalsy();
      expect(helper.toFileContent().content.trim()).toBe('bar;');
    });
  });

  describe(Helper.prototype.removeStatement, () => {
    it('should remove exsiting statement', () => {
      const helper = createSourceWriteHelper({ outputFileName: 'out.ts' });
      const st = ts.createExpressionStatement(ts.createIdentifier('hoge'));
      helper.pushStatement(st);
      expect(helper.removeStatement(st)).toBeTruthy();
      expect(helper.getStatements().length).toBe(0);
    });

    it('should return false when the statement is not found ', () => {
      const helper = createSourceWriteHelper({ outputFileName: 'out.ts' });
      const st = ts.createExpressionStatement(ts.createIdentifier('hoge'));
      helper.pushStatement(st);
      const st2 = ts.createExpressionStatement(ts.createIdentifier('hoge'));
      expect(helper.removeStatement(st2)).toBeFalsy();
      expect(helper.getStatements().length).toBe(1);
    });
  });

  describe(Helper.prototype.pushImportDeclaration, () => {
    it('should add statement at first when the helpper has no import statement', () => {
      const helper = createSourceWriteHelper({ outputFileName: 'out.ts' });
      helper.pushStatement(ts.createExpressionStatement(ts.createIdentifier('hoge')));
      helper.pushImportDeclaration(
        ts.createImportDeclaration(undefined, undefined, undefined, ts.createStringLiteral('typescript')),
      );
      expect(helper.toFileContent().content).toMatchSnapshot();
    });

    it('should add statement at next the last import declaration', () => {
      const helper = createSourceWriteHelper({ outputFileName: 'out.ts' });
      helper.pushStatement(
        ts.createImportDeclaration(undefined, undefined, undefined, ts.createStringLiteral('graphql')),
      );
      helper.pushStatement(ts.createExpressionStatement(ts.createIdentifier('hoge')));
      helper.pushImportDeclaration(
        ts.createImportDeclaration(undefined, undefined, undefined, ts.createStringLiteral('typescript')),
      );
      expect(helper.toFileContent().content).toMatchSnapshot();
    });
  });

  describe(Helper.prototype.pushNamedImportIfNeeded, () => {
    it('should not add import statement when the same statement exists', () => {
      const helper = createSourceWriteHelper({ outputFileName: 'out.ts' });
      expect(helper.pushNamedImportIfNeeded('Hoge', './foo')).toBeTruthy();
      expect(helper.pushNamedImportIfNeeded('Hoge', './foo')).toBeFalsy();
      expect(helper.getStatements().length).toBe(1);
      expect(helper.toFileContent().content).toMatchSnapshot();
    });

    it('should add import specifier if named import statement for same module already exists', () => {
      const helper = createSourceWriteHelper({ outputFileName: 'out.ts' });
      expect(helper.pushNamedImportIfNeeded('Foo', './foo')).toBeTruthy();
      expect(helper.pushNamedImportIfNeeded('Hoge', './foo')).toBeTruthy();
      expect(helper.getStatements().length).toBe(1);
      expect(helper.toFileContent().content).toMatchSnapshot();
    });
  });

  describe(Helper.prototype.pushDefaultImportIfNeeded, () => {
    it('should not add import statement when the same statement exists', () => {
      const helper = createSourceWriteHelper({ outputFileName: 'out.ts' });
      expect(helper.pushDefaultImportIfNeeded('Hoge', './foo')).toBeTruthy();
      expect(helper.pushDefaultImportIfNeeded('Hoge', './foo')).toBeFalsy();
      expect(helper.getStatements().length).toBe(1);
      expect(helper.toFileContent().content).toMatchSnapshot();
    });

    it('should add import specifier if named import statement for same module already exists', () => {
      const helper = createSourceWriteHelper({ outputFileName: 'out.ts' });
      expect(helper.pushNamedImportIfNeeded('Foo', './foo')).toBeTruthy();
      expect(helper.pushDefaultImportIfNeeded('Hoge', './foo')).toBeTruthy();
      expect(helper.getStatements().length).toBe(1);
      expect(helper.toFileContent().content).toMatchSnapshot();
    });
  });

  describe(Helper.prototype.writeLeadingComment, () => {
    it('should comment at the top of file', () => {
      const helper = createSourceWriteHelper({ outputFileName: 'out.ts' });
      helper.pushStatement(ts.createExpressionStatement(ts.createIdentifier('hoge')));
      helper.writeLeadingComment('foo');
      helper.writeLeadingComment('bar');
      expect(helper.toFileContent().content).toMatchSnapshot();
    });

    it('should comment when helper has no statement', () => {
      const helper = createSourceWriteHelper({ outputFileName: 'out.ts' });
      helper.writeLeadingComment('foo');
      helper.pushStatement(ts.createExpressionStatement(ts.createIdentifier('hoge')));
      expect(helper.toFileContent().content).toMatchSnapshot();
    });
  });
});
