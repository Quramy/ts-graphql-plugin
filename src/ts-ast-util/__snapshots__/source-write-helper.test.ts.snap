// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createSourceWriteHelper pushDefaultImportIfNeeded should add import specifier if named import statement for same module already exists 1`] = `
"import Hoge, { Foo } from \\"./foo\\";
"
`;

exports[`createSourceWriteHelper pushDefaultImportIfNeeded should not add import statement when the same statement exists 1`] = `
"import Hoge from \\"./foo\\";
"
`;

exports[`createSourceWriteHelper pushImportDeclaration should add statement at first when the helpper has no import statement 1`] = `
"import \\"typescript\\";
hoge;
"
`;

exports[`createSourceWriteHelper pushImportDeclaration should add statement at next the last import declaration 1`] = `
"import \\"typescript\\";
import \\"graphql\\";
hoge;
"
`;

exports[`createSourceWriteHelper pushNamedImportIfNeeded should add import specifier if named import statement for same module already exists 1`] = `
"import { Foo, Hoge } from \\"./foo\\";
"
`;

exports[`createSourceWriteHelper pushNamedImportIfNeeded should not add import statement when the same statement exists 1`] = `
"import { Hoge } from \\"./foo\\";
"
`;

exports[`createSourceWriteHelper writeLeadingComment should comment at the top of file 1`] = `
"/* foo */
/* bar */
hoge;
"
`;

exports[`createSourceWriteHelper writeLeadingComment should comment when helper has no statement 1`] = `
"/* foo */
hoge;
"
`;
